dart/common/EmbeddedAspect.h

```diff
@@ -0,0 +1,474 @@
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_EMBEDDEDASPECT_H_
+#define DART_COMMON_EMBEDDEDASPECT_H_
+
+#include "dart/common/detail/EmbeddedAspect.h"
+#include "dart/common/RequiresAspect.h"
+#include "dart/common/CompositeJoiner.h"
+
+namespace dart {
+namespace common {
+
+//==============================================================================
+/// This is the implementation of a standard embedded-state Aspect. Inherit
+/// the EmbedState class (next class down in the header) to use this Aspect
+/// implementation.
+///
+/// For more control over how your embedded-state Aspect is implemented, you can
+/// use the detail::EmbeddedStateAspect class.
+template <class CompositeT, typename StateDataT>
+class EmbeddedStateAspect : public detail::EmbeddedStateAspect<
+    CompositeTrackingAspect<CompositeT>,
+    EmbeddedStateAspect<CompositeT, StateDataT>, StateDataT>
+{
+public:
+
+  using Base = CompositeTrackingAspect<CompositeT>;
+  using Derived = EmbeddedStateAspect<CompositeT, StateDataT>;
+  using Impl = detail::EmbeddedStateAspect<Base, Derived, StateDataT>;
+  using State = typename Impl::State;
+  using StateData = typename Impl::StateData;
+
+  template <typename... Args>
+  EmbeddedStateAspect(Args&&... args)
+    : Impl(std::forward<Args>(args)...)
+  {
+    // Do nothing
+  }
+
+  virtual ~EmbeddedStateAspect() = default;
+
+};
+
+//==============================================================================
+/// Inherit this class to embed a State into your Composite object. DerivedT is
+/// the name of your class and StateDataT is a "plain-old data" structure that
+/// holds your state information.
+///
+/// Your derived class must implement the following functions:
+///
+/// \code{.cpp}
+/// void setAspectState(const AspectState& state);
+/// \endcode
+///
+/// To embed both state and properties information, use EmbedStateAndProperties.
+template <class DerivedT, typename StateDataT>
+class EmbedState : public virtual common::RequiresAspect<
```

Comment:
``Embed`` is being used as a verb while ``Requires`` is being used as an adjective.

``EmbedState`` is giving a command: "Embed this State"

``RequiresAspect`` is describing the derived class: "The derived class Requires this Aspect"